internal/integration/spec.go

```diff
@@ -0,0 +1,147 @@
+// Package integration ...
+// Copyright 2019 New Relic Corporation. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+package integration
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	yaml "gopkg.in/yaml.v2"
+)
+
+const fileNameMatcher = `^prometheus_.*\.ya?ml$`
+
+// Specs contains all the services specs mapped with the service name
+type Specs struct {
+	SpecsByName map[string]SpecDef
+}
+
+// SpecDef contains the rules to group metrics into entities
+type SpecDef struct {
+	Provider string      `yaml:"provider"`
+	Service  string      `yaml:"service"`
+	Entities []EntityDef `yaml:"entities"`
+}
+
+// EntityDef has info related to each entity
+type EntityDef struct {
+	Type       string        `yaml:"name"`
+	Properties PropertiesDef `yaml:"properties"`
+	Metrics    []MetricDef   `yaml:"metrics"`
+}
+
+// PropertiesDef defines the dimension used to get entity names
+type PropertiesDef struct {
+	Dimensions []string `yaml:"dimensions"`
+}
+
+// MetricDef contains metrics definitions
+type MetricDef struct {
+	Name string `yaml:"provider_name"`
+}
+
+// LoadSpecFiles loads all service spec files named like "prometheus_*.yml" that are in the filesPath
+func LoadSpecFiles(filesPath string) (Specs, error) {
+	specs := Specs{SpecsByName: make(map[string]SpecDef)}
+	var files []string
+
+	filesInPath, err := ioutil.ReadDir(filesPath)
+	if err != nil {
+		return specs, err
+	}
+	for _, f := range filesInPath {
+		if ok, _ := regexp.MatchString(fileNameMatcher, f.Name()); ok {
+			files = append(files, path.Join(filesPath, f.Name()))
+		}
+	}
+
+	for _, file := range files {
+		f, err := ioutil.ReadFile(file)
+		if err != nil {
+			logrus.Errorf("fail to read service spec file %s: %s ", file, err)
+			continue
+		}
+
+		var sd SpecDef
+		err = yaml.Unmarshal(f, &sd)
+		if err != nil {
+			logrus.Errorf("fail parse service spec file %s: %s", file, err)
+			continue
+		}
+		logrus.Debugf("spec file loaded for service:%s", sd.Service)
+		specs.SpecsByName[sd.Service] = sd
+	}
+
+	return specs, nil
+}
+
+// getEntity returns entity name and type of the metric based on the spec configuration defined for the service.
+// conditions for the metric:
+//   - metric.name has to start a prefix that matches with a service from the spec files
+//   - metric.name has to be defined in one of the entities of the spec file
+//   - if dimension has been specified for the entity, the metric need to have all of them.
+//   - metrics that belongs to entities with no dimension specified will share the same name
+func (s *Specs) getEntity(m Metric) (entityName string, entityType string, err error) {
```

Comment:
this is a helper function that will be used by the infra sdk emitter. metrics that return err will be added to the LocalEntity for now.

Reply:
Maybe for later, but if you expect that there will be different errors being returned, maybe we should have "typed" errors (like ErrServiceNotFound, ErrEntityNotFound, etc...)

Reply:
I was thinking not to use those errors more than for printing for debug proposes.